File: rnd_warehouse_management/public/js/work_order.js
// Work Order Form Enhancements for RND Warehouse Management

frappe.ui.form.on('Work Order', {
	// Form initialization
	refresh: function(frm) {
		// Add custom buttons
		add_work_order_buttons(frm);
		
		// Style zone status fields
		style_zone_status_fields(frm);
		
		// Show zone status dashboard
		show_zone_status_dashboard(frm);
	},
	
	// BOM change handler
	bom_no: function(frm) {
		if (frm.doc.bom_no) {
			update_material_requirements(frm);
		}
	},
	
	// Quantity change handler
	qty: function(frm) {
		if (frm.doc.bom_no) {
			update_material_requirements(frm);
		}
	}
});

// Helper Functions

function add_work_order_buttons(frm) {
	if (!frm.doc.__islocal) {
		// Add button to check material status
		frm.add_custom_button(__('Check Material Status'), function() {
			check_material_status(frm);
		}, __('Material Management'));
		
		// Add button to update zone status
		frm.add_custom_button(__('Update Zone Status'), function() {
			update_zone_status(frm);
		}, __('Material Management'));
		
		// Add button to create stock entry
		if (frm.doc.custom_current_zone_status === 'Green Zone') {
			frm.add_custom_button(__('Create Stock Entry'), function() {
				create_stock_entry_for_work_order(frm);
			}, __('Material Management'));
		}
		
		// Add button to view related stock entries
		frm.add_custom_button(__('View Stock Entries'), function() {
			view_related_stock_entries(frm);
		}, __('Material Management'));
	}
}

function style_zone_status_fields(frm) {
	if (frm.doc.custom_current_zone_status) {
		const zone_field = frm.get_field('custom_current_zone_status');
		if (zone_field) {
			const zone_wrapper = zone_field.$wrapper;
			
			// Remove existing classes
			zone_wrapper.removeClass('zone-red zone-green');
			
			// Add appropriate class
			if (frm.doc.custom_current_zone_status === 'Red Zone') {
				zone_wrapper.addClass('zone-red');
			} else if (frm.doc.custom_current_zone_status === 'Green Zone') {
				zone_wrapper.addClass('zone-green');
			}
		}
	}
	
	// Style completion percentage
	if (frm.doc.custom_material_completion_percentage !== undefined) {
		const completion_field = frm.get_field('custom_material_completion_percentage');
		if (completion_field) {
			const completion_wrapper = completion_field.$wrapper;
			const percentage = frm.doc.custom_material_completion_percentage;
			
			completion_wrapper.removeClass('completion-low completion-medium completion-high');
			
			if (percentage < 50) {
				completion_wrapper.addClass('completion-low');
			} else if (percentage < 100) {
				completion_wrapper.addClass('completion-medium');
			} else {
				completion_wrapper.addClass('completion-high');
			}
		}
	}
}

function show_zone_status_dashboard(frm) {
	if (frm.doc.custom_current_zone_status) {
		const status = frm.doc.custom_current_zone_status;
		const percentage = frm.doc.custom_material_completion_percentage || 0;
		const last_update = frm.doc.custom_last_zone_update;
		
		let color = status === 'Green Zone' ? 'green' : 'red';
		let message = `Zone Status: ${status} (${percentage.toFixed(1)}% materials available)`;
		
		if (last_update) {
			message += ` - Last updated: ${frappe.datetime.str_to_user(last_update)}`;
		}
		
		frm.dashboard.add_comment(message, color);
		
		// Add material shortage information if in Red Zone
		if (status === 'Red Zone' && frm.doc.custom_missing_materials_json) {
			try {
				const missing_materials = JSON.parse(frm.doc.custom_missing_materials_json);
				if (missing_materials.length > 0) {
					const shortage_message = `Missing ${missing_materials.length} material(s). Click 'Check Material Status' for details.`;
					frm.dashboard.add_comment(shortage_message, 'orange');
				}
			} catch (e) {
				// Ignore JSON parsing errors
			}
		}
	}
}

function check_material_status(frm) {
	frappe.call({
		method: 'rnd_warehouse_management.warehouse_management.work_order.get_work_order_material_status',
		args: {
			work_order_name: frm.doc.name
		},
		callback: function(r) {
			if (r.message && r.message.status === 'success') {
				show_material_status_dialog(r.message.material_status, r.message);
			} else {
				frappe.msgprint({
					title: __('Error'),
					message: r.message.message || __('Failed to get material status'),
					indicator: 'red'
				});
			}
		}
	});
}

function update_zone_status(frm) {
	frappe.call({
		method: 'rnd_warehouse_management.warehouse_management.work_order.update_work_order_zone_status',
		args: {
			work_order_name: frm.doc.name
		},
		callback: function(r) {
			if (r.message && r.message.status === 'success') {
				// Update form fields
				frm.set_value('custom_current_zone_status', r.message.zone_status);
				frm.set_value('custom_material_completion_percentage', r.message.completion_percentage);
				frm.set_value('custom_last_zone_update', r.message.last_updated);
				
				// Refresh displays
				style_zone_status_fields(frm);
				show_zone_status_dashboard(frm);
				
				frappe.show_alert({
					message: __('Zone status updated successfully'),
					indicator: 'green'
				}, 3);
			} else {
				frappe.msgprint({
					title: __('Error'),
					message: r.message.message || __('Failed to update zone status'),
					indicator: 'red'
				});
			}
		}
	});
}

function update_material_requirements(frm) {
	// This will be handled server-side when BOM or qty changes
	// Just refresh the zone status
	setTimeout(function() {
		update_zone_status(frm);
	}, 1000);
}

function create_stock_entry_for_work_order(frm) {
	const dialog = new frappe.ui.Dialog({
		title: __('Create Stock Entry'),
		fields: [
			{
				fieldname: 'purpose',
				label: __('Purpose'),
				fieldtype: 'Select',
				options: 'Material Issue\nMaterial Transfer',
				reqd: 1,
				default: 'Material Issue'
			},
			{
				fieldname: 'sap_movement_type',
				label: __('SAP Movement Type'),
				fieldtype: 'Select',
				options: '261 (FrontFlush - Goods Issue for Production)\n311 (BackFlush - Transfer for Kitting)',
				reqd: 1,
				default: '261'
			},
			{
				fieldname: 'qty',
				label: __('Quantity'),
				fieldtype: 'Float',
				default: frm.doc.qty
			}
		],
		primary_action_label: __('Create'),
		primary_action(values) {
			frappe.call({
				method: 'rnd_warehouse_management.warehouse_management.stock_entry.make_custom_stock_entry',
				args: {
					work_order: frm.doc.name,
					purpose: values.purpose,
					qty: values.qty
				},
				callback: function(r) {
					if (r.message) {
						// Set SAP movement type
						r.message.custom_sap_movement_type = values.sap_movement_type.split(' ')[0];
						
						// Open the new Stock Entry
						frappe.model.sync(r.message);
						frappe.set_route('Form', 'Stock Entry', r.message.name);
					}
				}
			});
			
			dialog.hide();
		}
	});
	
	dialog.show();
}

function view_related_stock_entries(frm) {
	frappe.route_options = {
		'custom_work_order_reference': frm.doc.name
	};
	frappe.set_route('List', 'Stock Entry');
}

function show_material_status_dialog(material_status, summary_data) {
	const dialog = new frappe.ui.Dialog({
		title: __('Material Status for Work Order'),
		size: 'large',
		fields: [
			{
				fieldname: 'summary_html',
				fieldtype: 'HTML'
			},
			{
				fieldname: 'material_html',
				fieldtype: 'HTML'
			}
		]
	});
	
	// Generate summary HTML
	let summary_html = '<div class="material-summary">';
	summary_html += `<h4>Summary</h4>`;
	summary_html += `<p><strong>Zone Status:</strong> <span class="${summary_data.zone_status === 'Green Zone' ? 'text-success' : 'text-danger'}">${summary_data.zone_status}</span></p>`;
	summary_html += `<p><strong>Completion:</strong> ${summary_data.completion_percentage.toFixed(1)}%</p>`;
	summary_html += '</div><hr>';
	
	// Generate detailed material HTML
	let material_html = '<h4>Detailed Material Status</h4>';
	material_html += '<table class="table table-bordered table-condensed">';
	material_html += '<thead><tr>';
	material_html += '<th>Item Code</th><th>Item Name</th><th>Required</th><th>Available</th><th>Shortage</th><th>Warehouse</th><th>Status</th>';
	material_html += '</tr></thead><tbody>';
	
	material_status.forEach(function(item) {
		const status_class = item.status === 'Available' ? 'text-success' : 'text-danger';
		const row_class = item.status === 'Available' ? '' : 'warning';
		
		material_html += `<tr class="${row_class}">`;
		material_html += `<td><strong>${item.item_code}</strong></td>`;
		material_html += `<td>${item.item_name}</td>`;
		material_html += `<td>${item.required_qty.toFixed(3)}</td>`;
		material_html += `<td>${item.available_qty.toFixed(3)}</td>`;
		material_html += `<td>${item.shortage.toFixed(3)}</td>`;
		material_html += `<td>${item.warehouse}</td>`;
		material_html += `<td class="${status_class}"><strong>${item.status}</strong></td>`;
		material_html += `</tr>`;
	});
	
	material_html += '</tbody></table>';
	
	dialog.fields_dict.summary_html.$wrapper.html(summary_html);
	dialog.fields_dict.material_html.$wrapper.html(material_html);
	dialog.show();
}

File: rnd_warehouse_management/rnd_warehouse_management/work_order.py
import frappe
from frappe import _
from frappe.utils import nowdate, now_datetime, flt

def before_save(doc, method=None):
	"""Hook: Before saving Work Order"""
	set_initial_zone_status(doc)
	update_material_requirements(doc)

def on_submit(doc, method=None):
	"""Hook: After submitting Work Order"""
	create_initial_material_assessment(doc)

def set_initial_zone_status(doc):
	"""Set initial zone status for new Work Orders"""
	if not doc.custom_current_zone_status:
		doc.custom_current_zone_status = "Red Zone"  # Default to Red Zone (incomplete)
		doc.custom_zone_status_color = "#dc3545"     # Red color
		doc.custom_material_completion_percentage = 0
		doc.custom_last_zone_update = now_datetime()

def update_material_requirements(doc):
	"""Update material requirements and availability"""
	if not doc.bom_no:
		return
	
	# Get BOM items
	bom_items = frappe.get_all(
		"BOM Item",
		filters={"parent": doc.bom_no},
		fields=["item_code", "qty", "warehouse", "item_name"]
	)
	
	# Calculate total required vs available
	total_items = len(bom_items)
	available_items = 0
	missing_materials = []
	
	for bom_item in bom_items:
		required_qty = flt(bom_item.qty * doc.qty)
		available_qty = get_available_qty(bom_item.item_code, bom_item.warehouse)
		
		if available_qty >= required_qty:
			available_items += 1
		else:
			missing_qty = required_qty - available_qty
			missing_materials.append({
				"item_code": bom_item.item_code,
				"item_name": bom_item.item_name,
				"required_qty": required_qty,
				"available_qty": available_qty,
				"missing_qty": missing_qty,
				"warehouse": bom_item.warehouse
			})
	
	# Update completion percentage
	completion_percentage = (available_items / total_items * 100) if total_items > 0 else 0
	doc.custom_material_completion_percentage = completion_percentage
	
	# Update zone status based on completion
	if completion_percentage >= 100:
		doc.custom_current_zone_status = "Green Zone"
		doc.custom_zone_status_color = "#28a745"  # Green
	else:
		doc.custom_current_zone_status = "Red Zone"
		doc.custom_zone_status_color = "#dc3545"  # Red
	
	# Store missing materials info
	doc.custom_missing_materials_json = frappe.as_json(missing_materials)

def get_available_qty(item_code, warehouse):
	"""Get available quantity for an item in a warehouse"""
	return frappe.db.get_value(
		"Bin",
		{"item_code": item_code, "warehouse": warehouse},
		"actual_qty"
	) or 0

def create_initial_material_assessment(doc):
	"""Create initial material assessment record"""
	# Create a material assessment log
	assessment = frappe.get_doc({
		"doctype": "Material Assessment Log",
		"work_order": doc.name,
		"assessment_date": now_datetime(),
		"zone_status": doc.custom_current_zone_status,
		"completion_percentage": doc.custom_material_completion_percentage,
		"missing_materials": doc.custom_missing_materials_json,
		"assessed_by": frappe.session.user
	})
	
	try:
		assessment.insert(ignore_permissions=True)
	except Exception as e:
		# If custom DocType doesn't exist, just log the assessment
		frappe.log_error(f"Material Assessment Log creation failed: {str(e)}")

def get_permission_query_conditions(user):
	"""Permission query conditions for Work Order"""
	if not user:
		user = frappe.session.user
	
	user_roles = frappe.get_roles(user)
	
	# Define role-based access
	role_access = {
		"Production Manager": "1=1",  # Access to all
		"Warehouse Manager": "1=1",   # Access to all
		"Manufacturing User": "(`tabWork Order`.owner = '{user}' OR `tabWork Order`.docstatus = 1)".format(user=user),
		"Stock User": "(`tabWork Order`.docstatus = 1)"  # Only submitted Work Orders
	}
	
	conditions = []
	for role in user_roles:
		if role in role_access:
			conditions.append(role_access[role])
	
	return " OR ".join(conditions) if conditions else "0=1"

@frappe.whitelist()
def update_work_order_zone_status(work_order_name):
	"""API endpoint to manually update Work Order zone status"""
	try:
		work_order = frappe.get_doc("Work Order", work_order_name)
		update_material_requirements(work_order)
		work_order.save(ignore_permissions=True)
		
		return {
			"status": "success",
			"zone_status": work_order.custom_current_zone_status,
			"completion_percentage": work_order.custom_material_completion_percentage,
			"last_updated": work_order.custom_last_zone_update
		}
	except Exception as e:
		frappe.log_error(f"Work Order zone status update failed: {str(e)}")
		return {"status": "error", "message": str(e)}

@frappe.whitelist()
def get_work_order_material_status(work_order_name):
	"""API endpoint to get detailed material status for a Work Order"""
	try:
		work_order = frappe.get_doc("Work Order", work_order_name)
		
		if not work_order.bom_no:
			return {"status": "error", "message": "No BOM associated with this Work Order"}
		
		# Get BOM items with availability
		bom_items = frappe.get_all(
			"BOM Item",
			filters={"parent": work_order.bom_no},
			fields=["item_code", "qty", "warehouse", "item_name", "description"]
		)
		
		material_status = []
		for bom_item in bom_items:
			required_qty = flt(bom_item.qty * work_order.qty)
			available_qty = get_available_qty(bom_item.item_code, bom_item.warehouse)
			
			material_status.append({
				"item_code": bom_item.item_code,
				"item_name": bom_item.item_name,
				"description": bom_item.description,
				"required_qty": required_qty,
				"available_qty": available_qty,
				"shortage": max(0, required_qty - available_qty),
				"warehouse": bom_item.warehouse,
				"status": "Available" if available_qty >= required_qty else "Shortage"
			})
		
		return {
			"status": "success",
			"work_order": work_order_name,
			"zone_status": work_order.custom_current_zone_status,
			"completion_percentage": work_order.custom_material_completion_percentage,
			"material_status": material_status
		}
	except Exception as e:
		frappe.log_error(f"Work Order material status check failed: {str(e)}")
		return {"status": "error", "message": str(e)}